### PR TITLE
Enable Gather gradient

### LINKED ISSUE #9669: Gather Op failing at back-prop with 3-D array and potentially causing JVM crash
#### Issue Description
- expected behavior : Gather Op works correctly for an array with any rank
- encountered behavior: This Op fails during back-prop (the part of the graph generated by org.nd4j.linalg.api.ops.impl.shape.Gather#doDiff()) during a Permute Op for a 3-D array. Batch size was 256. Last time reproduced with gradient NDArray having the following shape info:[3,256,256,768,196608,768,1,8192,1,99]. Gather dimension was 0, gather indices - [256,256], original array - matrix of shape [30522,768]. 

#### Version Information

* Deeplearning4j version - M2
* Platform information (OS, etc) - Windows 10 / Linux (Buster 10)
* CUDA version, if used - N/A
* NVIDIA driver version, if in use - N/A

#### Additional Information

The cause of the failure : Permute is being executed having only 2 dimensions instead of 3. Based on debug, I could identify that the root cause is incorrect calculation of dimension in org/nd4j/linalg/api/ops/impl/shape/Gather.java:182. Locally I used for some time an altered back-prop implementation which works fine. Because of a new load scheme for Custom Ops in M2 my custom implementation was ignored and this issue showed up.

Moreover, running the model on Linux caused intermittently a JVM crash. Tested with 3 different backends (`linux-x86_64`, `linux-x86_64-avx2` and `linux-x86_64-onednn-avx2`). The log of the crash is attached [here](https://gist.github.com/partarstu/3a101bce2dbb4e671c3d59e45c3fc00a) . After running the model with a custom back-prop implementation for Gather Op thus avoiding the issue described above, the JVM crash never happened again.

The only manifestation of an issue was the following error message on the console only on Windows platform (on Linux there was nothing) - 
`shape::doPermuteShapeInfo function failed: rearrange indexes are incorrect. Given permute indices must be < rank and >= 0.  Rearrange at index 2 was -1134333991`

_But the execution of the graph wasn't interrupted with this error, although the workflow was already broken._ This was probably what caused JVM crash at some point of the execution.

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
1. Fixes #9669  - the new gradient check passes old gradient checks
2. Add back missing code gen declarations for topk
3. Fix codegen to allow gather/gathernd to work with non numerical ndarrays (strings)
4. Adds new indexing feature allowing for recursive index resolution.
     a. This entailed adding a new recursive  indexing calls to get(..)
5. Add dynamic indexing resolution at runtime.
6. Adds new create view op allowing for views for proper in place views to be used.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

File: libnd4j/include/array/NDArray.hXX [hunk 1]
```diff
@@ -1201,6 +1201,22 @@ NDArray NDArray::varianceNumber(sd::variance::Ops op, bool biasCorrected) {
 }
 
 //////////////////////////////////////////////////////////////////////////
+
+
+
+NDArray NDArray::prodNumber() const {
+  if (isS()) throw std::runtime_error("NDArray::sumNumber: you can't use this method on String array!");
```
Comment (treo) [inline libnd4j/include/array/NDArray.hXX]: Wrong error message.

Comment (treo) [inline libnd4j/include/array/NDArray.hXX]: Wrong error message.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/ops/SDMath.java [hunk 2]
```diff
@@ -5500,7 +5500,7 @@ public SDVariable tanh(String name, SDVariable x) {
 
   /**
    * Matrix trace operation<br>
-   * For rank 2 matrices, the output is a scalar with the trace - i.e., sum of the main diagonal.<br>
+   * For rank 2 matrices, the output is a scalar vith the trace - i.e., sum of the main diagonal.<br>
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/ops/SDMath.java]: Reintroducing a typo that was previously fixed? 

Given that this is a generated namespace, I guess you'll need to fix the typo in the definition code.

Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/ops/SDMath.java]: Reintroducing a typo that was previously fixed? 

Given that this is a generated namespace, I guess you'll need to fix the typo in the definition code.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/custom/FusedBatchNorm.java [hunk 3]
```diff
@@ -97,4 +97,9 @@ public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
                 outputDataType == null ? DataType.FLOAT : outputDataType,
                 outputDataType == null ? DataType.FLOAT : outputDataType);
     }
+
+    @Override
+    public List<SDVariable> doDiff(List<SDVariable> f1) {
+        throw new UnsupportedOperationException("Automatic differentiation is not implemented!");
+    }
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/custom/FusedBatchNorm.java:104]: Huh? We don't have backprop for this?

Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/custom/FusedBatchNorm.java:104]: Huh? We don't have backprop for this?

**Review by treo — CHANGES_REQUESTED** [review]

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/custom/FusedBatchNorm.java (hunk 3 above)
Comment (agibsonccc) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/custom/FusedBatchNorm.java:104]: Nope

File: deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/resources/sentiment/sentiwordnet.txt [hunk 4]
```diff
@@ -69645,7 +69645,7 @@ n	09583506	0	0	thor#1	(Norse mythology) god of thunder and rain and farming; pic
 n	09583723	0	0	tyrr#1 tyr#1	(Norse mythology) god of war and strife and son of Odin; identified with Anglo-Saxon Tiu
 n	09583888	0.5	0	ullr#1 ull#1	(Norse mythology) one of the Aesir known for his beauty and skill with bow and skis; son of Sif and stepson of Thor
 n	09584079	0.125	0	vali#1	(Norse mythology) one of the Aesir and avenger of Balder; son of Odin
-n	09584218	0	0	vitharr#1 vithar#1 vidar#1	(Norse mythology) one of the Aesir; son of Odin; avenges his parent by slaying Fenrir at Ragnarok
+n	09584218	0	0	witharr#1 withar#1 vidar#1	(Norse mythology) one of the Aesir; son of Odin; avenges his parent by slaying Fenrir at Ragnarok
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/resources/sentiment/sentiwordnet.txt]: Now you've introduced new typos

**Review by treo — CHANGES_REQUESTED** [review]

File: deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/resources/sentiment/sentiwordnet.txt [hunk 5]
```diff
@@ -79810,7 +79810,7 @@ n	11348160	0	0	tree#3 sir_herbert_beerbohm_tree#1	English actor and theatrical p
 n	11348356	0	0	trevelyan#2 sir_george_otto_trevelyan#1 george_otto_trevelyan#1	English historian who wrote a history of the American revolution and a biography of his uncle Lord Macaulay (1838-1928)
 n	11348584	0	0	trevelyan#1 george_macaulay_trevelyan#1	English historian and son of Sir George Otto Trevelyan whose works include a social history of England and a biography of Garibaldi (1876-1962)
 n	11348812	0	0	trevino#1 supermex#1 lee_trevino#1 lee_buck_trevino#1	United States golfer (born in 1939)
-n	11348946	0	0	trevithick#1 richard_trevithick#1	English engineer who built the first railway locomotive (1771-1833)
+n	11348946	0	0	trewithick#1 richard_trewithick#1	English engineer who built the first railway locomotive (1771-1833)
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/resources/sentiment/sentiwordnet.txt]: This needs to be reverted too. I thought that marking just one instance in this file would have been enough.

**Review by treo — CHANGES_REQUESTED** [review]

File: libnd4j/include/ops/declarable/generic/tensor/create_view.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+#include <indexing/NDIndexUtils.h>
+#if NOT_EXCLUDED(OP_create_view)
+
+#include <ops/declarable/CustomOperations.h>
+
+namespace sd {
+namespace ops {
+
+CUSTOM_OP_IMPL(create_view, -2, -1, true, 0, -2) {
+  sd_printf("About to get variable 0\n",0);
```
Comment (quickwritereader) [inline libnd4j/include/ops/declarable/generic/tensor/create_view.cpp]: sd_debug

File: libnd4j/include/ops/declarable/generic/tensor/create_view.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  auto inputBase = INPUT_VARIABLE(0);
+  sd_printf("Obtained variable 0\n",0);
```
Comment (quickwritereader) [inline libnd4j/include/ops/declarable/generic/tensor/create_view.cpp]: sd_debug

File: libnd4j/include/array/impl/DataBuffer.cpp [hunk 8]
```diff
@@ -32,274 +32,276 @@ namespace sd {
...
-  _isOwnerPrimary = false;
-  _isOwnerSpecial = false;
-  _deviceId = sd::AffinityManager::currentDeviceId();
-
-  setCountersToZero();
-}
+    DataBuffer::DataBuffer() {
+        _primaryBuffer = nullptr;
+        _specialBuffer = nullptr;
+        _lenInBytes = 0;
```
Comment (quickwritereader) [inline libnd4j/include/array/impl/DataBuffer.cpp]: was this file formatted with the clangformat-style that comes in the libnd4j folder?

**Review by quickwritereader — DISMISSED** [review]

File: libnd4j/include/ops/declarable/generic/tensor/create_view.cpp (hunk 6 above)
Comment (agibsonccc) [reply libnd4j/include/ops/declarable/generic/tensor/create_view.cpp]: @quickwritereader I removed those afterwards thanks for checking!

File: libnd4j/include/array/impl/DataBuffer.cpp (hunk 8 above)
Comment (agibsonccc) [reply libnd4j/include/array/impl/DataBuffer.cpp]: done

**Review by quickwritereader — DISMISSED** [review]

File: libnd4j/include/ops/declarable/impl/DeclarableOp.cpp [hunk 9]
```diff
@@ -697,39 +698,75 @@ sd::Status sd::ops::DeclarableOp::execute(Context *block) {
...
 
-void DeclarableOp::overwriteResult(Context &block, int outputIdx, NDArray *array) {
-  throw std::runtime_error("Overwrite result used!");
-  // block.pushNDArrayToVariableSpace(block.nodeId(), outputIdx, array);
-  /*
-  auto varSpace = block.getVariableSpace();
-  if (varSpace->hasVariable(block.getNodeId(), outputIdx)) {
+
+void DeclarableOp::overwriteResult(Context &block, int outputIdx, NDArray *array,bool remove) {
+  sd_printf("Pushing variable\n",0);
```
Comment (treo) [inline libnd4j/include/ops/declarable/impl/DeclarableOp.cpp:703]: shouldn't this be an sd_debug? Same applies to other usages of sd_printf and printIndexedBuffer in this file

File: libnd4j/include/ops/declarable/impl/PlatformHelper.cpp [hunk 10]
```diff
@@ -76,8 +76,8 @@ sd::NDArray* PlatformHelper::getZ(graph::Context& ctx, int inputId) {
         sd_printf("Can't get Z variable for node_%i!\n", ctx.nodeId());
       }
     } else {
-      sd_printf("BOOM!\n", "");
-      throw std::runtime_error("Boom!");
+      sd_printf("Failed execution!\n", "");
+      throw std::runtime_error("Failed execution");
```
Comment (treo) [inline libnd4j/include/ops/declarable/impl/PlatformHelper.cpp]: While "Failed execution" is a better message than "Boom", I wonder if we can't have something more descriptive here?

Also, Do we need to both print and throw an exception?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Permute.java [hunk 11]
```diff
@@ -56,8 +56,9 @@ public Permute(INDArray input, INDArray result, int... permuteDims){
         addIArgument(permuteDims);
     }
 
-    public Permute(SameDiff sd, SDVariable input, SDVariable permuteDims){
+    public Permute(SameDiff sd, SDVariable input, SDVariable permuteDims) {
         super(sd, input, permuteDims);
+        System.out.println();
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Permute.java]: Leftover from debugging?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Slice.java [hunk 12]
```diff
@@ -42,27 +42,28 @@ public class Slice extends DynamicCustomOp {
...
         this.begin = begin;
         this.size = size;
         addIArgument(begin);
         addIArgument(size);
     }
 
-    public Slice(SameDiff sameDiff, @NonNull SDVariable input, @NonNull SDVariable begin, @NonNull SDVariable end){
+    public Slice(SameDiff sameDiff, @NonNull SDVariable input, @NonNull SDVariable begin, @NonNull SDVariable end) {
         super(null, sameDiff, new SDVariable[]{input, begin, end});
+        System.out.println();
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Slice.java]: Leftover from debugging?

**Review by treo — COMMENTED** [review]
Still need to read and consider the ADRs, but I've flagged some places for a clean up that I've found while given the code a cursory glance.

File: libnd4j/include/graph/impl/Context.cpp [hunk 13]
```diff
@@ -241,11 +241,14 @@ void Context::pushNDArrayToVariableSpace(std::pair<int, int> &pair, NDArray *arr
       _variableSpace->putVariable(pair, var);
       var->markRemovable(removable);
     } else {
+      sd_debug("Context: Getting variable in push ndarray",0);
       auto var = _variableSpace->getVariable(pair);
+      sd_printf("Context: After getting variable in push ndarray to variable space",0);
```
Comment (treo) [inline libnd4j/include/graph/impl/Context.cpp]: Still using sd_printf

File: libnd4j/include/graph/impl/Context.cpp [hunk 14]
```diff
@@ -293,6 +296,7 @@ Variable *Context::ensureVariable(int idx) {
     _variableSpace->putVariable(pair, var);
     return var;
   } else {
+    sd_printf("Before ensure variable",0);
```
Comment (treo) [inline libnd4j/include/graph/impl/Context.cpp]: Still using sd_printf

**Review by treo — CHANGES_REQUESTED** [review]
A bit more nitpicking

File: ADRs/0021 - Create View.md [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+1. Interval: INTERVAL_TYPE,2,1,start,end,stride,inclusive
+2. Point: POINT_TYPE,1,1,offset, DEFAULT_INCLUSIVE
+3. All: ALL_TYPE,0,1, DEFAULT_INCLUSIVE
+4. New Axis: NEW_AXIS,1,10, DEFAULT_INCLUSIVE
+
+
+Of note here are a few constants:
+1. *_TYPE: a pre defined constant representing the kind of index this is
+2. DEFAULT_INCLUSIVE: whether the index's end is inclusive or not (only needed for intervals)
+this is by default 0 most of the time since the value is only relevant for intervals.
```
Comment (treo) [inline ADRs/0021 - Create View.md]: Not quite sure what that means. If it is a constant, it should have the same value all of the time. 

Or is the above meant to represent the constructor signature for this op and here you're explaining what one of the argument values is?

File: ADRs/0021 - Create View.md [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+The format of the buffer is as follows:
+1. type of index
+2. Number of indices (representing offsets)
+3. Stride
+4. Inclusive/exclusive
+
+These are created as INT64 ndarrays passed in to the
+operation itself.
```
Comment (treo) [inline ADRs/0021 - Create View.md]: I think this should be explained a little bit better, in particular how this relates to the input representations above.

File: ADRs/0021 - Create View.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+Since SpecifiedIndex requires a copy most of the time, this op will
+mainly be focused on indexing that is guaranteed to use the same buffer.
+
+
+## Consequences
+
+### Advantages
+
+* Self contained op for creating a view
+* Similar in usage to indexes
```
Comment (treo) [inline ADRs/0021 - Create View.md:69]: Can you also explain how this view method differs from ND4J's? In particular what kinds of gotchas one might expect when switching between the two and how the API differs?

File: ADRs/0022 - Dynamic Indexing.md [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+## Proposal
+
+We introduce new dynamic indexing support in to nd4j's indexing engine.
+Previously, the indexing would not allow initialization with a negative index
+without passing in an ndarray. The problem with this is ndarrays
+are not known in samediff till execution time.
+
+We now support the ability to dynamically resolve indices
+during execution. This happens as follows:
+1. Each index now as a concept of initialized(). This returns a boolean indicating whether the operation was initialized or not. Initialized usually means there are no negative values present in the index.
```
Comment (treo) [inline ADRs/0022 - Dynamic Indexing.md]: > Initialized **usually** means there are no negative values present in the index

Can there be any cases when there still can be negative values present? Or does it actually **always** mean that there are no negative values present?

File: ADRs/0022 - Dynamic Indexing.md [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+2. If a negative index is specified, the index is set on the index itself but it will not be considered initialized.
+3. At runtime, upon use a deep copy of the index happens and then initialization will occur upon use relative to the ndarray using it.
```
Comment (treo) [inline ADRs/0022 - Dynamic Indexing.md]: Does this mean it is happening at the Java level? Or should this be done as late as possible and actually happen at the C++ level?

**Review by treo — CHANGES_REQUESTED** [review]
Overall I agree with the proposals in the ADRs, but I'd like to have them a bit more fleshed out. I've added some additional questions in the appropriate places. 

Also a note on on style: The ADRs are meant to be records of decisions that have been made. For this reason the proposal should be talking about things as if it is describing the current state of the code. 

This means "We introduce..." , "We now [support]..." and phrases that talk about things that can be considered future-ish, should be well considered. Usually there is a better way to phrase it.

Another thing is that in the Dynamic Indexing ADR, most of the first paragraph of the proposal should probably be part of the context instead, as you're not describing not the actual proposal but give more context on what was there before.

File: ADRs/0021 - Create View.md (hunk 15 above)
Comment (agibsonccc) [reply ADRs/0021 - Create View.md]: I explain that these are constant keys above. I clarified that now.

**Review by quickwritereader — DISMISSED** [review]

**Review by treo — APPROVED** [review]

**Pull request merged**